pandas/core/indexing.py

```diff
@@ -462,14 +471,20 @@ def _setitem_with_indexer(self, indexer, value):
             if isinstance(value, ABCSeries):
                 value = self._align_series(indexer, value)
 
-            info_idx = indexer[info_axis]
-            if is_integer(info_idx):
-                info_idx = [info_idx]
-            labels = item_labels[info_idx]
+            if has_missing_columns:
+                labels = [idx['key'] if isinstance(idx, dict) else
+                          item_labels[idx]
```

Comment:
@howsiwei you are adding WAY too much complexity here. You need to see the minimal change set to make this change work, it is likely way simpler.

Reply:
trace other setter calls and see

Reply:
@jreback mind suggesting where I should make the changes? The code for setting values is quite complicated and it's not obvious to me at all how to achieve the desired result with fewer changes.

Reply:
@jreback since `value` can have multiple types (eg. scalar, 1d array, 2d array, `DataFrame`), it's not easy to split `value` into value for each column. The only way I see to not duplicate the value splitting code is to modify the helper function `setter`, where `value` has already been split. I don't see how the code can be made much simpler when using this way.

Reply:
@jreback any updates?

Reply:
@jreback any suggestions?

Reply:
no suggestions as I have not looked at this recently
next week possibly

Reply:
@jreback sorry to disturb you once again. Have you looked into this PR?

Reply:
@howsiwei honestly i have not.

Reply:
so this is what I would do. I would touch no code here at all; rather add what you need into the loop above around 328. This handles the missing indexer case. 

Then we can assert that we have *no* missing items at all, IOW what you have at line 307 (what you define has_missing columns), could be true there but MUST be false at line 465.

This entire setting routine is *very* complex. We want to make this simpler and easier to understand; that may take a bigger refactor, but let's start by not adding additional complexity as much as possible.

i know this is non-trivial, but I am holding the line here because this is *already* pretty impenetrable code.

love for you to make it better!

Reply:
@jreback thanks for the suggestion. I'm guessing what you mean is that we could simply insert the column before setting. There are 2 downsides to this approach though:
- It may be less efficient.
- More importantly, currently `pandas` would try to infer the fill values for a column. Consider the following code:

``` python
import numpy as np
import pandas as pd

df = pd.DataFrame({'a': [0, 0]})
df.loc[0, 'b'] = pd.Timestamp('20120101')
df.loc[0, 'c'] = 1.0
print(df)
```

Output:
```
   a                    b    c
0  0  2012-01-01 00:00:00  1.0
1  0                  NaT  NaN
```

Note that column `b` and `c` are filled with `NaT` and `NaN` respectively.

To infer the fill values for each column, we need to split the values into separated columns first. But this splitting is only done after line 465.

Reply:
we already do this for expansions

my point is you are introducing another code path here

so either remove and fix the existing one (non trivial) or integrate this change

the dtype changes are expected in the current implementation

Reply:
> we already do this for expansions

Do what?

> the dtype changes are expected in the current implementation

What do you mean?

Reply:
@jreback Hmm after reading this again I think I understand what you said. Do you mean that using `NaN` as fill values for all columns is fine because this is already done elsewhere in pandas code? What are expansions though?

Reply:
@jreback can you please confirm if I should just fix this issue by inserting columns filled with `NaN` first?